README.markdown
Making time to manipulate time
==============================

The standard package for working with dates and times in Haskell, **time**, is
_awkward_. That's a subjective judgment, but over the years there have been few
areas more frustrating than trying to do pragmatic things with calendars and
clocks.

This package represents some opinionated approaches to working with times and
dates. It also is just a place to collect some hard-won idioms for converting
between things.

You absolutely do not need to use this; if you can make sense of the basic time
and calendar types available by default then by all means have at it.

This package was seeded using the TimeStamp type originally from the Vaultaire
project, a time-series database for systems metrics. That type was originally
implemented as a shim on top of base's **time** package; limitations there
(notably base being misleading about precision) have led it to be reimplemented
atop of the very complete (but also somewhat complicated) **hourglass**
library.

Our original use was wanting to conveniently measure things happening on
distributed computer systems. Since machine clock cycles are in units of
nanoseconds, this has the nice property that, assuming the system clock is not
corrupted, two subsequent events from the same source process are likely to
have monotonically increasing timestamps. And even if the system clock goes to
hell, they're still decently likely to be unique per device. Make for good
keys.

So the TimeStamp type herein is nanoseconds since the Unix epoch; which in
(signed) 64 bits means that you can represent times between early in the
morning of 21 September 1677 through just before midnight on 11 April 2262. The
primary use isn't doing calendaring, though; it's just working with machine
generated timestamps in distributed systems.

_Yes, at present that's all it is. A newtype over Int64, and some convenience
functions. Hopefully more common idioms will land here in due course._




CHANGELOG.markdown
* _v0.3.1_  
	Instances for to/from JSON objects using **aeson**.

* _v0.3.0_  
	Add Unbox instance to make TimeStamp suitable for **vector** operations

* _v0.2.1_  
	Initial public release


tests/TestSuite.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Test.Hspec

import CheckTimeStamp
import CheckVectorOperations

main :: IO ()
main = hspec suite

suite :: Spec
suite = do
    checkTimeStamp
    checkVectorUnboxed


tests/Experiment.hs
{-# LANGUAGE OverloadedStrings #-}
{-# OPTIONS -fno-warn-unused-imports #-}

module Main where

import Chrono.TimeStamp

main :: IO ()
main = do
    putStrLn "Hello"
 


tests/InstanceTest.hs
--
-- Data vault for metrics
--
-- Copyright © 2013-2014 Anchor Systems, Pty Ltd and Others
--
-- The code in this file, and the program it is a part of, is
-- made available to you by its authors as open source software:
-- you can redistribute it and/or modify it under the terms of
-- the 3-clause BSD licence.
--

-- | Test serialisation/deserialiastion for Vaultaire types.

import Test.Hspec
import Vaultaire.Types

main :: IO ()
main = hspec suite

suite :: Spec
suite =
  describe "Round trip through Read and Show instances" $ do
    it "outputs a correctly formated ISO 8601 timestamp when Shown" $ do
      show (TimeStamp 1406849015948797001) `shouldBe` "2014-07-31T23:23:35.948797001Z"
      show (TimeStamp 1406849015948797001) `shouldBe` "2014-07-31T23:23:35.948797001Z"
      show (TimeStamp 0) `shouldBe` "1970-01-01T00:00:00.000000000Z"

    it "Reads ISO 8601 timestamps" $ do
      read "2014-07-31T23:23:35.948797001Z" `shouldBe` TimeStamp 1406849015948797001
      read "2014-07-31T23:23:35Z" `shouldBe` TimeStamp 1406849015000000000
      read "2014-07-31" `shouldBe` TimeStamp 1406764800000000000

    it "reads the Unix epoch date" $
      read "1970-01-01" `shouldBe` TimeStamp 0

    it "permissively reads various formats" $ do
      show (read "1970-01-01T00:00:00.000000000Z" :: TimeStamp) `shouldBe` "1970-01-01T00:00:00.000000000Z"
      show (read "1970-01-01" :: TimeStamp) `shouldBe` "1970-01-01T00:00:00.000000000Z"
      show (read "0" :: TimeStamp) `shouldBe` "1970-01-01T00:00:00.000000000Z"


lib/Chrono/Compat.hs
--
-- Time to manipulate time
--
-- Copyright © 2013-2016 Operational Dynamics Consulting, Pty Ltd and Others
--
-- The code in this file, and the program it is a part of, is
-- made available to you by its authors as open source software:
-- you can redistribute it and/or modify it under the terms of
-- the 3-clause BSD licence.
--

--
-- | Compatibility with time types from other time handling libraries. Some of
-- these are just conveniences, but it's not always obvious how to convert
-- between time types even in the same package.
--
module Chrono.Compat
(
  -- * time package, from base
    convertToPosix
  , convertFromPosix
  , convertToUTC
  , convertFromUTC
  -- * hourglass package
  , convertToHourglass
  , convertFromHourglass
) where

import Data.Int (Int64)
import Data.Hourglass
import Data.Time.Clock
import Data.Time.Clock.POSIX
import Foreign.C.Types (CTime)

import Chrono.TimeStamp

--
-- | Utility function to convert nanoseconds since Unix epoch to a
-- 'NominalDiffTime', allowing you to then use the time manipulation
-- functions in "Data.Time.Clock" from __time__.
--
convertToPosix :: TimeStamp -> POSIXTime
convertToPosix = fromRational . (/ 1e9) . fromIntegral

--
--
-- | Utility function to convert nanoseconds since Unix epoch to a
-- 'ElapsedP', allowing you to then use the time manipulation
-- functions in the __hourglass__ package.
--
convertToHourglass :: TimeStamp -> ElapsedP
convertToHourglass = timeGetElapsedP

convertFromHourglass :: ElapsedP -> TimeStamp
convertFromHourglass = timeFromElapsedP


--
-- | Annoyingly, the various types in __time__ don't interoperate. Quite frequently
-- you need to get to, or from, 'UTCTime'.
--
convertToUTC :: TimeStamp -> UTCTime
convertToUTC x =
  let
    seconds :: TimeStamp -> CTime
    seconds = timeFromElapsedP . timeGetElapsedP
    utctime = posixSecondsToUTCTime . realToFrac . seconds
  in
    utctime x

convertFromUTC :: UTCTime -> TimeStamp
convertFromUTC = convertFromPosix . utcTimeToPOSIXSeconds

convertFromPosix :: POSIXTime -> TimeStamp
convertFromPosix =
  let
    nano :: POSIXTime -> Int64
    nano = floor . (* 1000000000) . toRational
  in
    TimeStamp . fromIntegral . nano


lib/Chrono/Formats.hs
--
-- Time to manipulate time
--
-- Copyright © 2013-2016 Operational Dynamics Consulting, Pty Ltd and Others
--
-- The code in this file, and the program it is a part of, is
-- made available to you by its authors as open source software:
-- you can redistribute it and/or modify it under the terms of
-- the 3-clause BSD licence.
--

{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module Chrono.Formats
(
      ISO8601_Precise(..)
    , ISO8601_Seconds(..)
    , ISO8601_Days(..)
    , Posix_Precise(..)
    , Posix_Micro(..)
    , Posix_Milli(..)
    , Posix_Seconds(..)
) where

import Data.Hourglass

--
-- | Format string describing full (nanosecond) precision ISO8601 time,
--
-- > 2014-07-31T23:09:35.274387019Z
--
data ISO8601_Precise = ISO8601_Precise

instance TimeFormat ISO8601_Precise where
    toFormat _ = TimeFormatString
        [ Format_Year
        , Format_Text '-'
        , Format_Month2
        , Format_Text '-'
        , Format_Day2
        , Format_Text 'T'
        , Format_Hour
        , Format_Text ':'
        , Format_Minute
        , Format_Text ':'
        , Format_Second
        , Format_Text '.'
        , Format_Precision 9
        , Format_Text 'Z'
        ]

data ISO8601_Seconds = ISO8601_Seconds

instance TimeFormat ISO8601_Seconds where
    toFormat _ = TimeFormatString
        [ Format_Year
        , Format_Text '-'
        , Format_Month2
        , Format_Text '-'
        , Format_Day2
        , Format_Text 'T'
        , Format_Hour
        , Format_Text ':'
        , Format_Minute
        , Format_Text ':'
        , Format_Second
        , Format_Text 'Z'
        ]

data ISO8601_Days = ISO8601_Days

instance TimeFormat ISO8601_Days where
    toFormat _ = TimeFormatString
        [ Format_Year
        , Format_Text '-'
        , Format_Month2
        , Format_Text '-'
        , Format_Day2
        ]

data Posix_Precise = Posix_Precise

instance TimeFormat Posix_Precise where
    toFormat _ = TimeFormatString
        [ Format_UnixSecond
        , Format_Text '.'
        , Format_MilliSecond
        , Format_MicroSecond
        , Format_NanoSecond
        ]

data Posix_Micro = Posix_Micro

instance TimeFormat Posix_Micro where
    toFormat _ = TimeFormatString
        [ Format_UnixSecond
        , Format_Text '.'
        , Format_MilliSecond
        , Format_MicroSecond
        ]

data Posix_Milli = Posix_Milli

instance TimeFormat Posix_Milli where
    toFormat _ = TimeFormatString
        [ Format_UnixSecond
        , Format_Text '.'
        , Format_MilliSecond
        ]

data Posix_Seconds = Posix_Seconds

instance TimeFormat Posix_Seconds where
    toFormat _ = TimeFormatString
        [ Format_UnixSecond
        ]


lib/Chrono/TimeStamp.hs
--
-- Time to manipulate time
--
-- Copyright © 2013-2017 Operational Dynamics Consulting, Pty Ltd and Others
--
-- The code in this file, and the program it is a part of, is
-- made available to you by its authors as open source software:
-- you can redistribute it and/or modify it under the terms of
-- the 3-clause BSD licence.
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE TypeFamilies #-}

module Chrono.TimeStamp
(
    TimeStamp(..)
  , getCurrentTimeNanoseconds

  , ISO8601_Precise(..)
) where

import Control.Applicative
import Data.Maybe
import Data.Int (Int64)
import Data.Hourglass
import qualified Data.Vector.Generic as G
import qualified Data.Vector.Generic.Mutable as M
import Data.Vector.Unboxed
import GHC.Generics
import Time.System

import Chrono.Formats

--
-- | Number of nanoseconds since the Unix epoch.
--
-- The Show instance displays the TimeStamp as seconds with the nanosecond
-- precision expressed as a decimal amount after the interger, ie:
--
-- >>> t <- getCurrentTimeNanoseconds
-- >>> show t
-- 2014-07-31T23:09:35.274387031Z
--
-- However this doesn't change the fact the underlying representation counts
-- nanoseconds since epoch:
--
-- >>> show $ unTimeStamp t
-- 1406848175274387031
--
-- There is a Read instance that is reasonably accommodating.
--
-- >>> read "2014-07-31T13:05:04.942089001Z" :: TimeStamp
-- 2014-07-31T13:05:04.942089001Z
--
-- >>> read "1406811904.942089001" :: TimeStamp
-- 2014-07-31T13:05:04.942089001Z
--
-- >>> read "1406811904" :: TimeStamp
-- 2014-07-31T13:05:04.000000000Z
--
-- In case you're wondering, the valid range of nanoseconds that fits into the
-- underlying Int64 is:
--
-- >>> show $ minBound :: TimeStamp
-- 1677-09-21T00:12:43.145224192Z
--
-- >>> show $ maxBound :: TimeStamp
-- 2262-04-11T23:47:16.854775807Z
--
-- so in a quarter millenium's time, yes, you'll have the Y2262 Problem.
-- Haskell code from today will, of course, still be running, so in the mid
-- Twenty-Third century you will need to replace this implementation with
-- something else.
--
newtype TimeStamp = TimeStamp {
    unTimeStamp :: Int64
} deriving (Eq, Ord, Enum, Num, Real, Integral, Bounded, Generic)

{-
    Hourglass works by sending types in and out of the Timeable and Time
    typeclasses. They're not particularly easy to work with, but they're a
    prerequisite for using timePrint
-}

instance Timeable TimeStamp where
    timeGetElapsedP :: TimeStamp -> ElapsedP
    timeGetElapsedP (TimeStamp ticks) =
      let
        (s,ns) = divMod ticks 1000000000
      in
        ElapsedP (Elapsed (Seconds (s))) (NanoSeconds (ns))

instance Time TimeStamp where
    timeFromElapsedP :: ElapsedP -> TimeStamp
    timeFromElapsedP (ElapsedP (Elapsed (Seconds seconds)) (NanoSeconds nanoseconds)) =
      let
        s  = fromIntegral seconds :: Int64
        ns = fromIntegral nanoseconds
      in
        TimeStamp $! (s * 1000000000) + ns


instance Show TimeStamp where
    show t = timePrint ISO8601_Precise t

instance Read TimeStamp where
    readsPrec _ s = maybeToList $ (,"") <$> parseInput s

parseInput :: String -> Maybe TimeStamp
parseInput = fmap reduceDateTime . parse
  where
    parse :: String -> Maybe DateTime
    parse x =
            timeParse ISO8601_Precise x
        <|> timeParse ISO8601_Seconds x
        <|> timeParse ISO8601_DateAndTime x -- from hourglass
        <|> timeParse ISO8601_Date x        -- from hourglass
        <|> timeParse Posix_Precise x
        <|> timeParse Posix_Micro x
        <|> timeParse Posix_Milli x
        <|> timeParse Posix_Seconds x

    reduceDateTime :: DateTime -> TimeStamp
    reduceDateTime = timeFromElapsedP . timeGetElapsedP

--
-- | Get the current system time, expressed as a 'TimeStamp' (which is to
-- say, number of nanoseconds since the Unix epoch).
--
getCurrentTimeNanoseconds :: IO TimeStamp
getCurrentTimeNanoseconds = do
    p <- timeCurrentP
    return $! convertToTimeStamp p

convertToTimeStamp :: ElapsedP -> TimeStamp
convertToTimeStamp = timeFromElapsedP

{-
    Ideally both the existing G.Vector Vector Int64 and M.MVector MVector
    Int64 instances would be sufficient to support automatically deriving
    those things for TimeStamp. But, *gaboom* somewhere down in GHC.Prim;
    writing this gumpf out manually did the trick, and satisifies the
    superclass requirements to derive an Unbox instance.
-}

newtype instance MVector s TimeStamp = MV_TimeStamp (MVector s Int64)
newtype instance Vector TimeStamp = V_TimeStamp (Vector Int64)

instance G.Vector Vector TimeStamp where
    basicUnsafeFreeze (MV_TimeStamp v) = V_TimeStamp <$> G.basicUnsafeFreeze v
    basicUnsafeThaw (V_TimeStamp v) = MV_TimeStamp <$> G.basicUnsafeThaw v
    basicLength (V_TimeStamp v) = G.basicLength v
    basicUnsafeSlice j k (V_TimeStamp v) = V_TimeStamp $ G.basicUnsafeSlice j k v
    basicUnsafeIndexM (V_TimeStamp v) i = TimeStamp <$> G.basicUnsafeIndexM v i
    {-# INLINE basicUnsafeFreeze #-}
    {-# INLINE basicUnsafeThaw #-}
    {-# INLINE basicLength #-}
    {-# INLINE basicUnsafeSlice #-}
    {-# INLINE basicUnsafeIndexM #-}

instance M.MVector MVector TimeStamp where
    basicLength (MV_TimeStamp v) = M.basicLength v
    basicUnsafeSlice j k (MV_TimeStamp v) = MV_TimeStamp $ M.basicUnsafeSlice j k v
    basicOverlaps (MV_TimeStamp a) (MV_TimeStamp b) = M.basicOverlaps a b
    basicUnsafeNew n = MV_TimeStamp <$> M.basicUnsafeNew n
    basicInitialize (MV_TimeStamp v) = M.basicInitialize v
    basicUnsafeRead (MV_TimeStamp v) i = TimeStamp <$> M.basicUnsafeRead v i
    basicUnsafeWrite (MV_TimeStamp v) i (TimeStamp a) = M.basicUnsafeWrite v i a
    {-# INLINE basicLength #-}
    {-# INLINE basicUnsafeSlice #-}
    {-# INLINE basicOverlaps #-}
    {-# INLINE basicUnsafeNew #-}
    {-# INLINE basicInitialize #-}
    {-# INLINE basicUnsafeRead #-}
    {-# INLINE basicUnsafeWrite #-}

deriving instance Unbox TimeStamp

